### ISSUE TITLE
When exporting MaskRCNN ONNX model from torchvision, the scale_factor is traced as the constant other than tensor

### ISSUE THREAD

**Issue by dxxz** [pytorch/pytorch#33443]
## 🐛 Bug
When exporting MaskRCNN ONNX model from torchvision, the scale_factor is traced as the constant other than tensor. So it makes the detection result of MaskRCNN ONNX model wrong with different size input image. 

## To Reproduce

1. Use PyTorch Nightly built version: pip install https://download.pytorch.org/whl/nightly/cu101/torch-1.5.0.dev20200216-cp36-cp36m-linux_x86_64.whl
1. In my local branch, I merge the workaround in this pull: [pytorch/vision#1840](https://github.com/pytorch/vision/pull/1840), and try with dynamic input size.
I upload the test script used in the local branch: https://gist.github.com/dxxz/3c40b5059bd5efe939ce96752d628158
1. python test_onnx_export.py --sanity-check

## Expected behavior
When running the MaskRCNN ONNX model with images of same input size, the detection results are right. But running on the images of different input sizem the detection results are wrong.

The logic of about code is using one image to export the Mask-RCNN ONNX model from pre-trained PyTorch model. Then run inference by Mask-RCNN ONNX model and PyTorch model individually with another sanity check image of different input sizes. Finally, compare the detection results by ONNX model and PyTorch model.

The compared results as follows:

    ONNXRuntime inputs:
    image
    [3, 'height', 'width']
    tensor(float)
    ONNXRuntime outputs:
    boxes
    ['obj', 4]
    tensor(float)
    labels
    ['obj']
    tensor(int64)
    scores
    ['obj']
    tensor(float)
    masks
    ['obj', 1, 'height', 'width']
    tensor(float)

    Sanity check the detection shape of pretrained model and ONNX model on the image for export process.
    0
    arr:     (100, 4)
    ort_arr: (100, 4)
    1
    arr:     (100,)
    ort_arr: (100,)
    2
    arr:     (100,)
    ort_arr: (100,)
    3
    arr:     (100, 1, 426, 640)
    ort_arr: (100, 1, 426, 640)

    Sanity check the detection shape of pretrained model and ONNX model on the sanity check images.
    0
    sanity_arr:     (1, 4)
    ort_sanity_arr: (4, 4)
    1
    sanity_arr:     (1,)
    ort_sanity_arr: (4,)
    2
    sanity_arr:     (1,)
    ort_sanity_arr: (4,)
    3
    sanity_arr:     (1, 1, 640, 586)
    ort_sanity_arr: (4, 1, 640, 586)

    Finish the sanity check process.

From this result, we can see the detection results by ONNX model and PyTorch model in the sanity check image are different.

The MaskRCNN ONNX export log as follows:

       %319 : Tensor = onnx::Shape(%318)
      %320 : Tensor = onnx::Constant[value={2}]()
      %321 : Long() = onnx::Gather[axis=0](%319, %320) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0
      %322 : Float() = onnx::Cast[to=1](%321) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0
      %323 : Float() = onnx::Constant[value={2.13333}]()
      %324 : Float() = onnx::Mul(%322, %323)
      %325 : Float() = onnx::Cast[to=1](%324) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0
      %326 : Float() = onnx::Floor(%325) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0
      %327 : Tensor = onnx::Shape(%318)
      %328 : Tensor = onnx::Constant[value={3}]()
      %329 : Long() = onnx::Gather[axis=0](%327, %328) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0
      %330 : Float() = onnx::Cast[to=1](%329) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0
      %331 : Float() = onnx::Constant[value={2.13333}]()
      %332 : Float() = onnx::Mul(%330, %331)
      %333 : Float() = onnx::Cast[to=1](%332) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0
      %334 : Float() = onnx::Floor(%333) # /scratch2/Chong_dxxz_Projects/Gitlab/MaskRCNN_ONNX/TorchVision/env/lib/python3.6/site-packages/torch/nn/functional.py:2538:0

The related function in torch/nn/functional.py:

    def _output_size(dim):
        _check_size_scale_factor(dim)
        if size is not None:
            return size
        scale_factors = _ntuple(dim)(scale_factor)
        # math.floor might return float in py2.7

        # make scale_factor a tensor in tracing so constant doesn't get baked in
        if torch._C._get_tracing_state():
            return [(torch.floor((input.size(i + 2).float() * torch.tensor(scale_factors[i],
                     dtype=torch.float32)).float())) for i in range(dim)]
        else:
            return [int(math.floor(float(input.size(i + 2)) * scale_factors[i])) for i in range(dim)]

functional.py:2538 is: 

            return [(torch.floor((input.size(i + 2).float() * torch.tensor(scale_factors[i],
                     dtype=torch.float32)).float())) for i in range(dim)]

From the log, we can see the scale_factors[i] be exported as the Constant, not the expected tensor.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5.0.dev20200216
 - OS (e.g., Linux): Ubuntu 16.04.6 LTS
 - How you installed PyTorch (`conda`, `pip`, source): pip install https://download.pytorch.org/whl/nightly/cu101/torch-1.5.0.dev20200216-cp36-cp36m-linux_x86_64.whl
 - Python version: 3.6
 - CUDA runtime version: 10.1.163
 - cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
 - GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
 - Any other relevant information:
[pip3] numpy==1.17.4
[pip3] torch==1.5.0.dev20200216
[pip3] torchvision==0.5.0a0+a4303c8
## Additional context

cc @suo @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by spandantiwari** [comment]
@lara-hdr - could you please take a look. Thanks.

**Comment by ankitdhall** [comment]
> When running the MaskRCNN ONNX model with images of same input size, the detection results are right. But running on the images of different input sizem the detection results are wrong.

hey @dxxz you mention that you are able to run `torchvision` MaskRCNN with ONNX. Could you share how exactly you converted it without facing any errors?

I face the issues documented in #33489 and was wondering how you got it up and running! :smile:

**Comment by dxxz** [comment]
@ankitdhall hi~~ I use the test script shown in the Reproduce Step 2 to export the MaskRCNN model.
Please note, you should merge [pytorch/vision#1840](https://github.com/pytorch/vision/pull/1840) to your local torchvision.

**Comment by dxxz** [comment]
Hi~~

If I hard-code the scale_factor as a constant float, then the results will be fine.
The result proves that this is the true issue need to be fixed to support dynamic input for MaskRCNN export.

    Sanity check the detection shape of pretrained model and ONNX model on the image for export process.
    0
    arr:     (15, 4)
    ort_arr: (15, 4)
    1
    arr:     (15,)
    ort_arr: (15,)
    2
    arr:     (15,)
    ort_arr: (15,)
    3
    arr:     (15, 1, 500, 375)
    ort_arr: (15, 1, 500, 375)

    Sanity check the detection shape of pretrained model and ONNX model on the sanity check images.
    0
    sanity_arr:     (17, 4)
    ort_sanity_arr: (17, 4)
    1
    sanity_arr:     (17,)
    ort_sanity_arr: (17,)
    2
    sanity_arr:     (17,)
    ort_sanity_arr: (17,)
    3
    sanity_arr:     (17, 1, 640, 428)
    ort_sanity_arr: (17, 1, 640, 428)

@spandantiwari @lara-hdr

**Comment by ankitdhall** [comment]
thanks @dxxz for the tip! it works by merging the https://github.com/pytorch/vision/pull/1840 to the local :smile: are you also converting it to TensorRT to accelerate performance for the MaskRCNN?

**Comment by dxxz** [comment]
@ankitdhall Yes, in the plan. Do you meet the constant scale_factor issue locally?

**Comment by ankitdhall** [comment]
@dxxz It is also in my plan to convert it to TensorRT to gain considerable speed ups during inference. I get the same error as you if I pass a different image size in.

**Comment by engineer1109** [comment]
where is your mask rcnn model from?

**Comment by xsidneib** [comment]
I have the same problem. The maskRNN is from the torchvision models.

``
>>> model = torchvision.models.detection.keypointrcnn_resnet50_fpn(pretrained=True)
>>> model.eval()
>>> x = [torch.rand(3, 300, 400), torch.rand(3, 500, 400)]
>>> predictions = model(x)
>>>
>>> # optionally, if you want to export the model to ONNX:
>>> torch.onnx.export(model, x, "keypoint_rcnn.onnx", opset_version = 11)

**Comment by neginraoof** [comment]
Thanks for reporting this issue. Many problems with export of Mask RCNN and Keypoint RCNN models with dynamic size detections and keypoints has been resolved recently. The main PR which addresses this issue is: https://github.com/pytorch/vision/pull/2087
And the latest one is: https://github.com/pytorch/vision/pull/2488
Could you please let us know if you're still facing this issue using the latest updates?

Thanks.

**Comment by spandantiwari** [comment]
@dxxz @ankitdhall @xsidneib - as @neginraoof asked, can one of you please check with the latest versions (which have the PRs  to see if you still see this issue. Otherwise, we can go ahead and close it.

**Comment by xsidneib** [comment]
Thanks @spandantiwari  and @neerajprad ,


I've got the nightly build yesterday and - in my case -  I could export the ONNX model, without problem. However, my issue is even more complicated and I actually don't know if the issue is with pytorch or TensorRT from NVIDIA. Basically, if I export the maskrcnn (as it's in the torchvision documentatino) to ONNX and try to convert it in TensorRT, I have issues which says:


While parsing node number 223 [Pad]:
ERROR: /home/jenkins/workspace/OSS/L0_MergeRequest/oss/parsers/onnx/builtin_op_importers.cpp:2219 In function importPad:
[8] Assertion failed: inputs.at(1).is_weights()
[09/23/2020-08:41:42] [E] Failed to parse onnx file


Unfortunately,  my lack of knowledge in this area doesn't help me to get more information what exactly is failing. Just looks like the ONNX file isn't compatible somehow.


Do you have any experience with TensorRT? I would really appreciate if we could investigate it and have it working end to end, but it might is an issue in the TensorRT side. I need more details to ask it in their forum.


https://github.com/onnx/onnx-tensorrt/issues/411 following the https://github.com/NVIDIA/TensorRT/issues/439 .. It looks they are already in touch with PyTorch on that.


Thanks a lot

**Comment by spandantiwari** [comment]
@xsidneib - thanks for getting back. Good to know that you can export the model now. I too am not sure about this error that you have included. Looks like this is some issue when importing the ONNX model in TensorRT. From a quick look at the TensorRT issues that you have listed, they don't seem to be questioning the correctness of the exported ONNX model, but discussing the additional nodes that are in the graph in opset 11. Let me know if I got that wrong.

As you suggest already, could you please followup on the TensorRT forum about this?

Since you can export the model now, I will close this issue now. Please feel free to reopen if you see this issue again.
